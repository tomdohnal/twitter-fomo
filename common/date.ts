import dayjs, { Dayjs } from 'dayjs';
import utc from 'dayjs/plugin/utc';

dayjs.extend(utc);

export const dayjsUtc = dayjs.utc;

export const DAY_NOW = dayjsUtc().startOf('day');
export const DAY_BEFORE_ONE_WEEK = DAY_NOW.subtract(1, 'week');

export { Dayjs };
