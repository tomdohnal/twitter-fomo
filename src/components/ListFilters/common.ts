import { Filters } from '../../filters';

export const PERIOD_ITEMS = [
  {
    label: 'Day',
    value: 'DAY',
  },
  {
    label: 'Week',
    value: 'WEEK',
  },
];

export const COMMUNITY_ITEMS = ['React', 'Vue', 'Angular'];

export const TWEET_TYPE_ITEMS = [
  {
    label: 'Text',
    value: 'TEXT',
  },
  {
    label: 'Media',
    value: 'MEDIA',
  },
  {
    label: 'Link',
    value: 'LINK',
  },
];

export const ACCOUNT_TYPE_ITEMS = [
  {
    label: 'Personal',
    value: 'PERSONAL',
  },
  {
    label: 'Page',
    value: 'BUSINESS',
  },
];

interface FilterAction {
  value: Filters[keyof Filters];
  name: keyof Filters;
}

export const createGetNewFilters = (filters: Filters) => (action: FilterAction): Filters => {
  const oldField = filters[action.name];
  const newField = Array.isArray(oldField)
    ? oldField.includes(action.value as string)
      ? oldField.filter((value) => value !== action.value)
      : [...oldField, action.value]
    : action.value;

  return { ...filters, [action.name]: newField };
};

export const INITIAL_FILTERS: Filters = {
  period: 'DAY',
  accountTypes: [],
  communities: [],
  tweetTypes: [],
};

export interface Props {
  filters: Filters;
  setFilters(filters: Filters): void;
  setFilters(fn: (filters: Filters) => Filters): void;
}
