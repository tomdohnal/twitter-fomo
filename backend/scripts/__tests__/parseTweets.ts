import * as R from 'ramda';
import {
  createTweetTypeFilters,
  createCommunitiesFilters,
  AccountTweet,
  createDateFilters,
} from '../../filters';
import {
  createTweet,
  createCommunity,
  createAccount,
  createTextTweet,
  createLinkTweet,
  createMediaTweet,
} from '../../testUtils';
import { dayjsUtc } from '../../../common/date';
import { createListsData } from '../parseTweets';

describe('parseTweets', () => {
  describe('createListInputs', () => {
    it('works with NO filters', () => {
      const sortedAccountTweets: AccountTweet[] = [
        {
          account: createAccount(),
          tweets: [
            createTweet({ created_at: '2020-01-01' }),
            createTweet({ created_at: '2020-01-01' }),
            createTweet({ created_at: '2020-01-01' }),
          ],
        },
      ];

      const listObjects = createListsData({
        sortedAccountTweets,
        appliedFilters: [],
        remainingFilters: [],
      });

      expect(listObjects).toHaveLength(1);
      expect(listObjects).toMatchSnapshot();
    });

    it('works with DATE filters', async () => {
      // 2020-01-01
      const tweet1 = createTweet({ created_at: '2020-01-01' });
      const tweet2 = createTweet({ created_at: '2020-01-01' });
      const tweet3 = createTweet({ created_at: '2020-01-01' });
      const tweet4 = createTweet({ created_at: '2020-01-01' });

      // 2020-01-02
      const tweet5 = createTweet({ created_at: '2020-01-02' });
      const tweet6 = createTweet({ created_at: '2020-01-02' });

      const sortedAccountTweets: AccountTweet[] = [
        {
          account: createAccount(),
          tweets: [tweet1, tweet2, tweet5],
        },
        {
          account: createAccount(),
          tweets: [tweet3, tweet4, tweet6],
        },
      ];

      const dateFilters = createDateFilters(dayjsUtc('2020-01-01').add(1, 'day'));

      const listObjects = createListsData({
        sortedAccountTweets,
        appliedFilters: [],
        remainingFilters: [dateFilters],
      });

      expect(listObjects).toHaveLength(3); // 1 without filters + 2 for each filter (1 day filter + 1 week filter)
      expect(listObjects).toMatchSnapshot();
    });

    it('works with DATE and COMMUNITIES and TWEET_TYPE filters', () => {
      const COMMUNITIES = [createCommunity(), createCommunity()];
      const account1 = createAccount({
        communities: [COMMUNITIES[0]],
      });

      const account2 = createAccount({
        communities: [COMMUNITIES[1]],
      });

      // 2020-01-01
      const tweet1 = createTweet({ created_at: '2020-01-01' });
      const tweet2 = createTextTweet({ created_at: '2020-01-01' });
      const tweet3 = createLinkTweet({ created_at: '2020-01-01' });
      const tweet4 = createMediaTweet({ created_at: '2020-01-01' });

      // 2020-01-02
      const tweet5 = createTweet({ created_at: '2020-01-02' });
      const tweet6 = createTweet({ created_at: '2020-01-02' });

      const sortedAccountTweets: AccountTweet[] = [
        {
          account: account1,
          tweets: [tweet1, tweet2, tweet5],
        },
        {
          account: account2,
          tweets: [tweet3, tweet4, tweet6],
        },
      ];

      // yields 2 filters - 1 DAY filters and 1 WEEK filter
      const dateFilters = createDateFilters(dayjsUtc('2020-01-01').add(1, 'day'));

      // yields 2 filters
      const communitiesFilters = createCommunitiesFilters(COMMUNITIES);

      // yields 3 filters (TEXT, LINK and MEDIA)
      const tweetTypeFilters = createTweetTypeFilters();

      const listObjects = createListsData({
        sortedAccountTweets,
        appliedFilters: [],
        remainingFilters: [dateFilters, communitiesFilters, tweetTypeFilters],
      });

      expect(listObjects).toHaveLength(36); // draw a tree... (it's hard for me as well, don't worry, LOL :D)
      expect(listObjects.map(R.omit(['tweets']))).toMatchSnapshot();
    });
  });
});
