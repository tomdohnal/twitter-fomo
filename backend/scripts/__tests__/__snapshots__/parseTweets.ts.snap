// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseTweets createListInputs works with DATE and COMMUNITIES and TWEET_TYPE filters 1`] = `
Array [
  Object {},
  Object {
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "LINK",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "MEDIA",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "TEXT",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "LINK",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "MEDIA",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "TEXT",
  },
  Object {
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "LINK",
  },
  Object {
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "MEDIA",
  },
  Object {
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "TEXT",
  },
  Object {
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "LINK",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "MEDIA",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "TEXT",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "LINK",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "MEDIA",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "TEXT",
  },
  Object {
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "LINK",
  },
  Object {
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "MEDIA",
  },
  Object {
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweetType": "TEXT",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "tweetType": "LINK",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "tweetType": "MEDIA",
  },
  Object {
    "community": Object {
      "connect": "1",
    },
    "tweetType": "TEXT",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "tweetType": "LINK",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "tweetType": "MEDIA",
  },
  Object {
    "community": Object {
      "connect": "2",
    },
    "tweetType": "TEXT",
  },
  Object {
    "tweetType": "LINK",
  },
  Object {
    "tweetType": "MEDIA",
  },
  Object {
    "tweetType": "TEXT",
  },
]
`;

exports[`parseTweets createListInputs works with DATE filters 1`] = `
Array [
  Object {
    "tweets": Object {
      "create": Array [
        Object {
          "account": Object {
            "connect": "412801ae-3e44-4790-9992-f94b3186fb76",
          },
          "accountName": "Alex_McGlynn54",
          "favoritesCount": 90337,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Ab voluptas sed quae nam eos sint.",
          "twitterId": "94459",
        },
        Object {
          "account": Object {
            "connect": "ec940c92-5980-4eb5-bec9-704edb8f1272",
          },
          "accountName": "Therese_Beatty12",
          "favoritesCount": 76292,
          "publishedAt": "2020-01-02T00:00:00.000Z",
          "text": "Recusandae consectetur optio laudantium et incidunt possimus molestias.",
          "twitterId": "20856",
        },
        Object {
          "account": Object {
            "connect": "8eded1d0-420e-4a19-aaf6-83b452acd600",
          },
          "accountName": "Isai.Koss29",
          "favoritesCount": 69232,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Fuga doloribus distinctio placeat ullam minima ducimus temporibus modi.",
          "twitterId": "7336",
        },
        Object {
          "account": Object {
            "connect": "fbe024f2-3162-465a-ae8d-65a837e308ae",
          },
          "accountName": "Michel.Wisoky9",
          "favoritesCount": 41702,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Blanditiis natus et eos itaque velit.",
          "twitterId": "80074",
        },
        Object {
          "account": Object {
            "connect": "06010831-d483-48fd-8134-199f6850dc43",
          },
          "accountName": "Bridie15",
          "favoritesCount": 17227,
          "publishedAt": "2020-01-02T00:00:00.000Z",
          "text": "Aut laborum iure molestiae et facere placeat molestiae.",
          "twitterId": "42838",
        },
        Object {
          "account": Object {
            "connect": "72ec6682-ae55-49ca-b2eb-9ac55c45ec6e",
          },
          "accountName": "Trycia.Jaskolski",
          "favoritesCount": 13747,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Aut nam et dolorum architecto fugit repellendus hic veritatis qui.",
          "twitterId": "57838",
        },
      ],
    },
  },
  Object {
    "period": "WEEK",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweets": Object {
      "create": Array [
        Object {
          "account": Object {
            "connect": "412801ae-3e44-4790-9992-f94b3186fb76",
          },
          "accountName": "Alex_McGlynn54",
          "favoritesCount": 90337,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Ab voluptas sed quae nam eos sint.",
          "twitterId": "94459",
        },
        Object {
          "account": Object {
            "connect": "ec940c92-5980-4eb5-bec9-704edb8f1272",
          },
          "accountName": "Therese_Beatty12",
          "favoritesCount": 76292,
          "publishedAt": "2020-01-02T00:00:00.000Z",
          "text": "Recusandae consectetur optio laudantium et incidunt possimus molestias.",
          "twitterId": "20856",
        },
        Object {
          "account": Object {
            "connect": "8eded1d0-420e-4a19-aaf6-83b452acd600",
          },
          "accountName": "Isai.Koss29",
          "favoritesCount": 69232,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Fuga doloribus distinctio placeat ullam minima ducimus temporibus modi.",
          "twitterId": "7336",
        },
        Object {
          "account": Object {
            "connect": "fbe024f2-3162-465a-ae8d-65a837e308ae",
          },
          "accountName": "Michel.Wisoky9",
          "favoritesCount": 41702,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Blanditiis natus et eos itaque velit.",
          "twitterId": "80074",
        },
        Object {
          "account": Object {
            "connect": "06010831-d483-48fd-8134-199f6850dc43",
          },
          "accountName": "Bridie15",
          "favoritesCount": 17227,
          "publishedAt": "2020-01-02T00:00:00.000Z",
          "text": "Aut laborum iure molestiae et facere placeat molestiae.",
          "twitterId": "42838",
        },
        Object {
          "account": Object {
            "connect": "72ec6682-ae55-49ca-b2eb-9ac55c45ec6e",
          },
          "accountName": "Trycia.Jaskolski",
          "favoritesCount": 13747,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Aut nam et dolorum architecto fugit repellendus hic veritatis qui.",
          "twitterId": "57838",
        },
      ],
    },
  },
  Object {
    "period": "DAY",
    "startDate": "2020-01-01T00:00:00.000Z",
    "tweets": Object {
      "create": Array [
        Object {
          "account": Object {
            "connect": "412801ae-3e44-4790-9992-f94b3186fb76",
          },
          "accountName": "Alex_McGlynn54",
          "favoritesCount": 90337,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Ab voluptas sed quae nam eos sint.",
          "twitterId": "94459",
        },
        Object {
          "account": Object {
            "connect": "8eded1d0-420e-4a19-aaf6-83b452acd600",
          },
          "accountName": "Isai.Koss29",
          "favoritesCount": 69232,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Fuga doloribus distinctio placeat ullam minima ducimus temporibus modi.",
          "twitterId": "7336",
        },
        Object {
          "account": Object {
            "connect": "fbe024f2-3162-465a-ae8d-65a837e308ae",
          },
          "accountName": "Michel.Wisoky9",
          "favoritesCount": 41702,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Blanditiis natus et eos itaque velit.",
          "twitterId": "80074",
        },
        Object {
          "account": Object {
            "connect": "72ec6682-ae55-49ca-b2eb-9ac55c45ec6e",
          },
          "accountName": "Trycia.Jaskolski",
          "favoritesCount": 13747,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Aut nam et dolorum architecto fugit repellendus hic veritatis qui.",
          "twitterId": "57838",
        },
      ],
    },
  },
  Object {
    "period": "DAY",
    "startDate": "2020-01-02T00:00:00.000Z",
    "tweets": Object {
      "create": Array [
        Object {
          "account": Object {
            "connect": "ec940c92-5980-4eb5-bec9-704edb8f1272",
          },
          "accountName": "Therese_Beatty12",
          "favoritesCount": 76292,
          "publishedAt": "2020-01-02T00:00:00.000Z",
          "text": "Recusandae consectetur optio laudantium et incidunt possimus molestias.",
          "twitterId": "20856",
        },
        Object {
          "account": Object {
            "connect": "06010831-d483-48fd-8134-199f6850dc43",
          },
          "accountName": "Bridie15",
          "favoritesCount": 17227,
          "publishedAt": "2020-01-02T00:00:00.000Z",
          "text": "Aut laborum iure molestiae et facere placeat molestiae.",
          "twitterId": "42838",
        },
      ],
    },
  },
]
`;

exports[`parseTweets createListInputs works with NO filters 1`] = `
Array [
  Object {
    "tweets": Object {
      "create": Array [
        Object {
          "account": Object {
            "connect": "ec940c92-5980-4eb5-bec9-704edb8f1272",
          },
          "accountName": "Therese_Beatty12",
          "favoritesCount": 76292,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Recusandae consectetur optio laudantium et incidunt possimus molestias.",
          "twitterId": "20856",
        },
        Object {
          "account": Object {
            "connect": "01ae3e44-7909-4992-b94b-3186fb7680c8",
          },
          "accountName": "Alex_McGlynn54",
          "favoritesCount": 52479,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Nam eos sint.",
          "twitterId": "94459",
        },
        Object {
          "account": Object {
            "connect": "72ec6682-ae55-49ca-b2eb-9ac55c45ec6e",
          },
          "accountName": "Trycia.Jaskolski",
          "favoritesCount": 13747,
          "publishedAt": "2020-01-01T00:00:00.000Z",
          "text": "Aut nam et dolorum architecto fugit repellendus hic veritatis qui.",
          "twitterId": "57838",
        },
      ],
    },
  },
]
`;
